### ISSUE TITLE
Feature request: Check if any import matches

### ISSUE THREAD

**Issue by hzaun** [sverweij/dependency-cruiser#328]
### Context

Our codebase is moving to a new arch and I want to add a warning if a module hasn't been migrated yet. Essentially, I want to check if a type of dependency exists.

Let's say I want to check if all the controller modules have been migrated, which I can simplify into a check to see if they extend the new base controller, for eg: 

`class 'AbcController' extends 'GenericController'`

### Expected Behavior

I want to check if AbcController contains an import for GenericController:

```
    {
      name: 'no-clean-arch-controller',
      severity: 'warn',
      "from": { "path": "^src\/api\/([^/]+)\/([a-zA-Z]+Controller.ts)"},
      "to": { 
        "pathNot": "src\/framework\/GenericController.ts" 
      },
    }
```
with the result:

```
no-clean-arch-controller: src/api/x/abcController.ts → src/framework/GenericController.ts
```

### Current Behavior

It currently tells me about all the imports that aren't GenericControllerImpl, and nothing

```
no-clean-arch-controller: src/api/x/abcController.ts → src/example/some-import.ts
no-clean-arch-controller: src/api/x/abcController.ts → src/example/another-import.ts
```

### Possible Solution

A boolean to signify that there's no dependency from A to B would be helpful.

### Considered alternatives

I tried to make it work with _allowed_, but it became too complex and didn't solve my issue anyway.

**Comment by sverweij** [comment]
Hi @hzaun thanks for this feature request! It's an interesting use case. It might be possible with the existing types of rules , but if so that'll likely only be in a complicated rule set - as you already discovered. And that while your need is reasonable, and seems simple and straightforward (correct me if I'm wrong b.t.w.): every _controller_ must depend on the _GenericController_.

So dependency-cruiser will have to grow a feature for this. Besides your suggestion I'm toying with the idea of adding a '_required_' rule section on top of the two existing _forbidden_ and _allowed_ ones (see below - could that work for you?). I'll try to have a closer look during the weekend.

```javascript
{
  forbidden: [
    // bunch of forbidden rules
  ],
  required: [
    {
      name: 'no-clean-arch-controller',
      description: 'Every controller is expected to inherit from the Generic controller', 
      severity: 'warn',
      from: { "path": "^src\/api\/([^/]+)\/([a-zA-Z]+Controller.ts)"},
      to: { 
        path: "src\/framework\/GenericController.ts" 
      },
    },
    // potentially more required rules
  ],
}

**Comment by hzaun** [comment]
@sverweij That would be perfect! (:

**Comment by sverweij** [comment]
hi @hzaun -  I've cobbled together a version of dependency-cruiser that validates against rules in `required` section. It's published on npm as a beta (`dependency-cruiser@9.9.0-beta-1`). It still needs polishing (doc, more expansive tests) but it should cover your use case.

=> Could you let me know whether this is fit for your purpose (e.g. what things you still miss, which stuff didn't work as you expected)?

[dependency-cruiser-repro-repo/328](https://github.com/sverweij/dependency-cruiser-repro-repo/tree/main/328) contains a working example.

**Comment by sverweij** [comment]
Just a ping: I've done some polishing. To make it more future proof I've renamed the 'from' attribute to 'module' so in the future it's possible to use the 'from' attribute for other useful purposes. E.g. this one to make sure that each source file has an associated spec file in the unit tests folder that directly points to it:

```javascript
{
  required: [{
    name: "spec-per-source",
    comment: 'for each source file there should be at least a corresponding spec file in the test folder',
    module: { path: "^src/(.+)\\.js$"},
    from: { path: "^test/$1\\.spec\\.js$"}
  }]
}
```

**Comment by hzaun** [comment]
Hey! I've tried it out- it's just what I needed! I would like a modification if possible..

Currently, if a 'required' rule isn't fulfilled, I get a warning for _all_ the imports it has, which ends up polluting the result.
Eg. I have 21 controllers that don't extend the generic controller, and now I have 200 violations just from the 'no-clean-arch-controller' rule.

**Comment by sverweij** [comment]
Hi @hzaun thanks for the quick feedback! 

Sorry you bumped into an issue! Which version did you use?

I did have the same problem in beta-2 and beta-3. In 9.9.0-beta-4 it should've been fixed, though (tested in the [repro repo](https://github.com/sverweij/dependency-cruiser-repro-repo/tree/main/328)) - so if you used dependency-cruiser@9.9.0-beta-4 and still have the issue I'll need some additional information to reproduce it (i.e. the configuration file)...

**Comment by hzaun** [comment]
I was using dependency-cruiser@9.9.0-beta-3, just updated to beta-4, it works as expected!

**Comment by hzaun** [comment]
I was using dependency-cruiser@9.9.0-beta-3, just updated to beta-4, it works as expected, thank you so much!

**Comment by sverweij** [comment]
As you can see I've merged the PR - this feature will be in dependency-cruiser's next regular release (will appear within a few days). I'll give a ping in this thread when that happens.

**Comment by sverweij** [comment]
@hzaun I've just published dependency-cruiser@9.9.0 which includes the merged PR - enjoy :-) .
